### PR TITLE
[JENKIN-57795] Orphan nodes logic

### PR DESCRIPTION
**Pull request by pyieh**
Addresses issue of orphan nodes not being re-attached when the plugin has already hit an instance cap. Because the orphan re-attachment logic was tied to the EC2Cloud.provision() logic, if the instance cap was hit, then provisioning wasn't even attempted and the orphan re-attachment isn't triggered.

Add functionality to EC2Cloud to attempt to re-attach orphan/stopped nodes even when its unable to provision more nodes due to hitting the instance cap.

### CODE REVIEW

**Comment by res0nance** [comment]
The test seems to be failing

**Comment by pyieh** [comment]
The tests were passing for me locally. I've added some more ignores as per a suggestion online to see if that clears up the failing PowerMock test.

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java [hunk 1]
```diff
@@ -645,6 +627,21 @@ else if (jenkinsInstance.isTerminating()) {
         }
     }
 
+    private void attachSlavesToJenkins(Jenkins jenkins, List<EC2AbstractSlave> slaves, SlaveTemplate t) throws IOException {
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:630]:
```suggestion
    private static void attachSlavesToJenkins(Jenkins jenkins, List<EC2AbstractSlave> slaves, SlaveTemplate t) throws IOException {
```

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 2]
```diff
@@ -18,61 +18,56 @@
...
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.servlet.ServletException;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:66]: Could you not use star imports?

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 3]
```diff
@@ -729,7 +712,12 @@ private void logProvisionInfo(String message) {
             diFilters.add(new Filter("availability-zone").withValues(getZone()));
         }
 
-        String subnetId = chooseSubnetId();
+        String subnetId;
+        if (rotateSubnet) {
+            subnetId = chooseSubnetId();
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:717]: can choose subnet id handle this specific?

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 4]
```diff
@@ -740,19 +728,20 @@ private void logProvisionInfo(String message) {
             /*
              * If we have a subnet ID then we can only use VPC security groups
              */
-            if (!securityGroupSet.isEmpty()) {
-                List<String> groupIds = getEc2SecurityGroups(ec2);
+            List<String> groupIds = null;
+            if (!getSecurityGroupSet().isEmpty()) {
+                groupIds = getEc2SecurityGroups(ec2);
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:733]:
```suggestion
            if (!getSecurityGroupSet().isEmpty()) {
                List<String> groupIds = getEc2SecurityGroups(ec2);
```

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 5]
```diff
@@ -768,6 +757,44 @@ private void logProvisionInfo(String message) {
...
+        List<TagSpecification> tagList = new ArrayList<>();
+        TagSpecification tagSpecification = new TagSpecification();
+        tagSpecification.setTags(instTags);
+        tagList.add(tagSpecification.clone().withResourceType(ResourceType.Instance));
+        tagList.add(tagSpecification.clone().withResourceType(ResourceType.Volume));
+        riRequest.setTagSpecifications(tagList);
+
+        HashMap<RunInstancesRequest, List<Filter>> ret = new HashMap<>();
+        ret.put(riRequest, diFilters);
+        return ret;
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:773]:
```suggestion
        return Collections.singletonMap(riRequest, diFilters);
```

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 6]
```diff
@@ -768,6 +816,42 @@ private void logProvisionInfo(String message) {
...
+        }
+
+        List<TagSpecification> tagList = new ArrayList<>();
+        TagSpecification tagSpecification = new TagSpecification();
+        tagSpecification.setTags(instTags);
+        tagList.add(tagSpecification.clone().withResourceType(ResourceType.Instance));
+        tagList.add(tagSpecification.clone().withResourceType(ResourceType.Volume));
+        riRequest.setTagSpecifications(tagList);
+
+        return (HashMap<RunInstancesRequest, List<Filter>>) Collections.singletonMap(riRequest, diFilters);
```
[the remark this answers is not in the document]
Comment (pyieh) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: Without the cast, there's a type check complaint. I've removed using Collections.singletonMap and just reverted it to using the HashMap with a single entry.

**Comment by pyieh** [comment]
@res0nance there seems to be some upper bound dependencies checks failing. Can you explain how those can be fixed? They seem to be coming from other dependencies that this project doesn't manage?

**Comment by res0nance** [comment]
Looks like this was caused by a pipeline library issue (It's failing on master as well), I'll work on fixing it

**Comment by pyieh** [comment]
@res0nance Wondering if there's been any update on this? It seems to be failing with a different error now.

**Comment by res0nance** [comment]
I've fixed the CI on our part it looks like an infra issue right now

**Comment by jetersen** [comment]
I wrote this piece of groovy script to remove orphaned disconnected agents.
Would this PR solve this issue?

```groovy
import jenkins.model.Jenkins
import hudson.plugins.ec2.EC2AbstractSlave

println "Cleaning up offline slaves..."
Jenkins.instance.slaves.each { node ->
  if(!(node in EC2AbstractSlave)) return
  def computer = node.getComputer()
  if (computer && !computer.isOffline()) return
  def logs = computer.getLog()
  if (!logs) return
  if (!"Connection terminated".equals(logs.readLines().last())) return
  println "Deleting ${node.name}"
  computer.doDoDelete()
}

null
```

**Comment by jetersen** [comment]
link to see current commit author: https://github.com/jenkinsci/ec2-plugin/commit/fc6528041fa2d6958a265ba96ef73ab7c8d1165d.patch

@pyieh 
You might want to set up an email alias on your GitHub account 🥇
https://help.github.com/articles/adding-an-email-address-to-your-github-account/

Or look into fixing your git user.email config 😅  
https://help.github.com/articles/setting-your-commit-email-address-in-git
then you would need to amend your commits: https://stackoverflow.com/a/3042512
after changing author

to get proper credit for your commits 🏆

**Comment by pyieh** [comment]
@jetersen yeah I realized I was pushing to my personal repo using my work email. Honestly I'm not that concerned with "proper credit". It does show under my name as the committer, albeit not tied to my Github account so no worries.

**Comment by pyieh** [comment]
@res0nance if everything looks good, can we get this merged?

**Comment by pyieh** [comment]
@res0nance Wanted to follow up and see if we can get this merged. Thank you

**Comment by pyieh** [comment]
@fcojfernandez Can I get my PR merged? @res0nance approved it a while back, but it hasn't been merged yet.

**Comment by mbaranczak** [comment]
A big thank-you to pyieh for fixing this annoying bug.

I see that the patch has been merged into master, but there hasn't been an official release yet. Does anybody know when it will be released?

**Comment by manojtr** [comment]
Hello Folks, any update on the release with this fix?

**Comment by res0nance** [comment]
This has been released since `1.51` you can see the release notes here https://github.com/jenkinsci/ec2-plugin/releases

**Comment by manojtr** [comment]
Today I tried 1.53 version and the issue is not resolved. 

```
SlaveTemplate{ami='ami-038e073abe89730b3', labels='win2016dlp'}. Attempting to provision slave needed by excess workload of 1 units
Nov 04, 2020 12:36:01 PM INFO hudson.plugins.ec2.EC2Cloud getNewOrExistingAvailableSlaveSlaveTemplate{ami='ami-038e073abe89730b3', labels='win2016dlp'}. Cannot provision - no capacity for instances: 0
Nov 04, 2020 12:36:01 PM WARNING hudson.plugins.ec2.EC2Cloud provisionCan't raise nodes for SlaveTemplate{ami='ami-038e073abe89730b3', labels='win2016dlp'}
```
However, Jenkins identified the node and the node details screen shows "Launch Agent" Button. But the agent is not running. 

Please note, this is a windows agent.

**Comment by manojtr** [comment]
> link to see current commit author: https://github.com/jenkinsci/ec2-plugin/commit/fc6528041fa2d6958a265ba96ef73ab7c8d1165d.patch
> 
> @pyieh
> You might want to set up an email alias on your GitHub account 🥇
> https://help.github.com/articles/adding-an-email-address-to-your-github-account/
> 
> Or look into fixing your git user.email config 😅
> https://help.github.com/articles/setting-your-commit-email-address-in-git
> then you would need to amend your commits: https://stackoverflow.com/a/3042512
> after changing author
> 
> to get proper credit for your commits 🏆

@jetersen how did you use this script? As an init script or a scheduled job running in a very short interval? Since the official fix is not working, I am looking to implement something like this. Thank you

**Comment by manojtr** [comment]
> I wrote this piece of groovy script to remove orphaned disconnected agents.
> Would this PR solve this issue?
> 
> Specifally seeing EC2 agent on Linux that will stay connected for a good while and than suddenly get stuck on `Connection terminated` and never get cleaned up.
> 
> ```groovy
> import jenkins.model.Jenkins
> import hudson.plugins.ec2.EC2AbstractSlave
> 
> println "Cleaning up offline EC2 slaves..."
> Jenkins.instance.slaves.each { node ->
>   if(!(node in EC2AbstractSlave)) return
>   def computer = node.getComputer()
>   if (computer && !computer.isOffline()) return
>   def logs = computer.getLog()
>   if (!logs) return
>   if (!"Connection terminated".equals(logs.readLines().last())) return
>   println "Deleting ${node.name}"
>   computer.doDoDelete()
> }
> 
> null
> ```

@jetersen how did you use this script? As an init script or a scheduled job running in a very short interval? Since the official fix is not working, I am looking to implement something like this. Thank you

**Comment by jetersen** [comment]
As a pipeline job that would curl the script console and use a jenkins token from credentials

**Comment by manojtr** [comment]
Thanks @jetersen for the quick reply.

**Comment by szubster** [comment]
The added method `attemptReattachOrphanOrStoppedNodes` is not used anywhere besides tests. This PR does nothing?

**Pull request merged**